src/fast_graph32.rs

```diff
@@ -0,0 +1,218 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+use std::convert::TryFrom;
+
+use serde::Deserialize;
+use serde::Serialize;
+
+use crate::fast_graph::FastGraphEdge;
+use crate::FastGraph;
+
+/// Special graph data-structure that is identical to `FastGraph` except that it uses u32 integers
+/// instead of usize integers. This is used to store a `FastGraph` in a 32bit representation on disk
+/// when using a 64bit system.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct FastGraph32 {
+    num_nodes: u32,
+    pub ranks: Vec<u32>,
+    pub edges_fwd: Vec<FastGraphEdge32>,
+    pub first_edge_ids_fwd: Vec<u32>,
+
+    pub edges_bwd: Vec<FastGraphEdge32>,
+    pub first_edge_ids_bwd: Vec<u32>,
+}
+
+impl FastGraph32 {
+    /// Creates a 32bit Graph from a given `FastGraph`. All (potentially 64bit) `usize` integers are
+    /// simply converted to u32 and if a value exceeds the 32bit limit an error is thrown. The only
+    /// exception is `std::u32::MAX`, which is converted to `std::usize::MAX`.
+    pub fn new(fast_graph: &FastGraph) -> Self {
+        FastGraph32 {
+            num_nodes: usize_to_u32(fast_graph.get_num_nodes()),
+            ranks: usize_to_u32_vec(&fast_graph.ranks),
+            edges_fwd: usize_to_u32_edges(&fast_graph.edges_fwd),
+            first_edge_ids_fwd: usize_to_u32_vec(&fast_graph.first_edge_ids_fwd),
+            edges_bwd: usize_to_u32_edges(&fast_graph.edges_bwd),
+            first_edge_ids_bwd: usize_to_u32_vec(&fast_graph.first_edge_ids_bwd),
+        }
+    }
+
+    /// Converts a 32bit Graph to an actual `FastGraph` using `usize` such that it can be used with
+    /// FastPaths crate. Any integers that equal `std::u32::MAX` are mapped to `std::usize::MAX`.
+    pub fn convert_to_usize(self) -> FastGraph {
+        let mut g = FastGraph::new(self.num_nodes as usize);
+        g.ranks = u32_to_usize_vec(&self.ranks);
+        g.edges_fwd = u32_to_usize_edges(&self.edges_fwd);
+        g.first_edge_ids_fwd = u32_to_usize_vec(&self.first_edge_ids_fwd);
+        g.edges_bwd = u32_to_usize_edges(&self.edges_bwd);
+        g.first_edge_ids_bwd = u32_to_usize_vec(&self.first_edge_ids_bwd);
+        return g;
+    }
+}
+
+/// 32bit equivalent to `FastGraphEdge`, see `FastGraph32` docs.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct FastGraphEdge32 {
+    pub base_node: u32,
+    pub adj_node: u32,
+    pub weight: u32,
+    pub replaced_in_edge: u32,
+    pub replaced_out_edge: u32,
+}
+
+fn usize_to_u32(int: usize) -> u32 {
+    if int.eq(&std::usize::MAX) {
+        return usize_to_u32(std::u32::MAX as usize);
+    }
+    return u32::try_from(int)
+        .expect(format!("Could not convert {} to a 32bit integer", int).as_str());
```

Comment:
This is probably fine (if the benchmarks aren't affected by this change), but it does eagerly evaluate the string every single time, which might actually add up to lots for the all the usizes being converted here. Not sure if it gets compiled away or not. https://stackoverflow.com/questions/44922542/is-it-possible-to-implement-a-method-that-takes-a-format-string has some possible patterns; personally I think the most straightforward is just:

`if let Ok(x) = u32::try_from(int) {
  x
} else {
  panic!("Could not convert {} to a 32-bit integer", int);
}`

Reply:
Ok thanks. Btw is simply using `panic` fine for you here or would you prefer passing an (error)result to the caller of `save_to_disk32`?